presto-hive/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java

```diff
@@ -257,4 +268,46 @@ public static void verifyCanDropColumn(ExtendedHiveMetastore metastore, String d
             throw new PrestoException(NOT_SUPPORTED, "Cannot drop the only non-partition column in a table");
         }
     }
+
+    public static FileSystem getFileSystem(HdfsEnvironment hdfsEnvironment, HdfsContext context, Path path)
+    {
+        try {
+            return hdfsEnvironment.getFileSystem(context, path);
+        }
+        catch (IOException e) {
+            throw new PrestoException(HIVE_FILESYSTEM_ERROR, format("Error getting file system. Path: %s", path), e);
+        }
+    }
+
+    public static void renameFile(FileSystem fileSystem, Path source, Path target, Supplier<Boolean> cancelled)
+    {
+        if (cancelled.get()) {
+            return;
+        }
+        try {
+            if (fileSystem.exists(target) || !fileSystem.rename(source, target)) {
+                throw new PrestoException(HIVE_FILESYSTEM_ERROR, getRenameErrorMessage(source, target));
+            }
+        }
+        catch (IOException e) {
+            throw new PrestoException(HIVE_FILESYSTEM_ERROR, getRenameErrorMessage(source, target), e);
+        }
+    }
+
+    private static String getRenameErrorMessage(Path source, Path target)
+    {
+        return format("Error moving data files from %s to final location %s", source, target);
+    }
+
+    public static void waitForListenableFutures(List<ListenableFuture<?>> listenableFutures)
```

Comment:
Is this the right way to wait for a list of ListenableFuture? -- Googled but didn't find a definitive answer  @nezihyigitbasi , @arhimondr

Reply:
You can use `Futures.whenAllSucceed`. 

That method returns a single `ListenableFuture` that you can wait on. It has the fail fast semantics. As long as any underlying future failed - the future returnes by the `Futures.whenAllSucceed`.

Just double check the cancellation semantics in case of a failure. Check that underlying futures are being cancelled when one of the underlying futures fails.

Reply:
Thanks! I checked `AggregateFuture#afterDone`, and it seems if the aggregate future is cancelled, it will also cancel all the underlying futures.